datafusion/expr/src/logical_plan/mutate.rs

```diff
@@ -0,0 +1,346 @@
+            LogicalPlan::Join(Join { left, right, .. }) => {
+                rewrite_arc(left, &mut f)?.and_then(|| rewrite_arc(right, &mut f))
+            }
+            LogicalPlan::CrossJoin(CrossJoin { left, right, .. }) => {
+                rewrite_arc(left, &mut f)?.and_then(|| rewrite_arc(right, &mut f))
+            }
+            LogicalPlan::Limit(Limit { input, .. }) => rewrite_arc(input, &mut f),
+            LogicalPlan::Subquery(Subquery { subquery, .. }) => {
+                rewrite_arc(subquery, &mut f)
+            }
+            LogicalPlan::SubqueryAlias(SubqueryAlias { input, .. }) => {
+                rewrite_arc(input, &mut f)
+            }
+            LogicalPlan::Extension(extension) => {
+                rewrite_extension_inputs(&mut extension.node, &mut f)
+            }
+            LogicalPlan::Union(Union { inputs, .. }) => inputs
+                .iter_mut()
+                .try_fold(Transformed::no(()), |acc, input| {
+                    acc.and_then(|| rewrite_arc(input, &mut f))
+                }),
+            LogicalPlan::Distinct(
+                Distinct::All(input) | Distinct::On(DistinctOn { input, .. }),
+            ) => rewrite_arc(input, &mut f),
+            LogicalPlan::Explain(explain) => rewrite_arc(&mut explain.plan, &mut f),
+            LogicalPlan::Analyze(analyze) => rewrite_arc(&mut analyze.input, &mut f),
+            LogicalPlan::Dml(write) => rewrite_arc(&mut write.input, &mut f),
+            LogicalPlan::Copy(copy) => rewrite_arc(&mut copy.input, &mut f),
+            LogicalPlan::Ddl(ddl) => {
+                if let Some(input) = ddl.input_mut() {
+                    rewrite_arc(input, &mut f)
+                } else {
+                    Ok(Transformed::no(()))
+                }
+            }
+            LogicalPlan::Unnest(Unnest { input, .. }) => rewrite_arc(input, &mut f),
+            LogicalPlan::Prepare(Prepare { input, .. }) => rewrite_arc(input, &mut f),
+            LogicalPlan::RecursiveQuery(RecursiveQuery {
+                static_term,
+                recursive_term,
+                ..
+            }) => rewrite_arc(static_term, &mut f)?
+                .and_then(|| rewrite_arc(recursive_term, &mut f)),
+            // plans without inputs
+            LogicalPlan::TableScan { .. }
+            | LogicalPlan::Statement { .. }
+            | LogicalPlan::EmptyRelation { .. }
+            | LogicalPlan::Values { .. }
+            | LogicalPlan::DescribeTable(_) => Ok(Transformed::no(())),
+        }?;
+
+        // after visiting the actual children we we need to visit any subqueries
+        // that are inside the expressions
+        children_result.and_then(|| self.rewrite_subqueries(&mut f))
+    }
+
+    /// applies `f` to LogicalPlans in any subquery expressions
+    ///
+    /// If Err is returned, the plan may be left in a partially modified state
+    fn rewrite_subqueries<F>(&mut self, mut f: F) -> Result<Transformed<()>>
```

Comment:
I'm a bit concerned about this part. I mean the `TreeNode` APIs are currently inconsistent about handling "inner" trees. I'm refferring to `LogicalPlan` for example. It overrides `TreeNode::apply()` and `TreeNode::visit()` to apply `f` on subquery's logical plans too.
- But first of all, none of the other trees do that,
- secodly, `LogicalPlan` doesn't override `rewrite()` or `transform...()` so transformations are applied only the "current" tree and not on "whole" logical plan.

I'm not sure that the current state is right or wrong as I can imagine both usescases in which
- a transformations needs to be applied on the current tree only
- a transformations need to change the whole tree (subplans trees included)

Now, this code seems to apply `f` on subquery logical plans too. If we keep it that way `mutate()` will be inconsistent with other APIs. But I'm also not sure that why it applies `f` on the root node only? Shouldn't it call `subquery.subquery.mutate(f)` to start a traversal on the subquery plan tree?

Is it needed because of the `Optimizer`? Does it optimize the whole logical plan at once? If so, I think we should introduce new APIs on `LogicalPlan` that is different to the existing `TreeNode` APIs. In Spark for example we have `TreeNode.transform()` and `QueryPlan.transformWithSubqueries()` for the different purposes.

Reply:
Shall I open a PR that adds `LogicalPlan::rewriteWithSubqueries()` and `LogicalPlan::transform...WithSubqueries()`?

Reply:
> Now, this code seems to apply f on subquery logical plans too. If we keep it that way mutate() will be inconsistent with other APIs

I actually found I needed to handle recursing into subqueries to get some optimizer tests tests passing. Interestingly, recursing into subqueries also seems to have improved planning in some cases like https://github.com/apache/arrow-datafusion/pull/9780#discussion_r1545618832


> Is it needed because of the Optimizer? Does it optimize the whole logical plan at once? If so, I think we should introduce new APIs on LogicalPlan that is different to the existing TreeNode APIs. In Spark for example we have TreeNode.transform() and QueryPlan.transformWithSubqueries() for the different purposes.

`OptimizerRule`s can currently specify if they want to handle recursion themselves or not: https://docs.rs/datafusion/latest/datafusion/optimizer/trait.OptimizerRule.html#method.apply_order


> I'm not sure that the current state is right or wrong as I can imagine both usescases in which
> 
> * a transformations needs to be applied on the current tree only
> * a transformations need to change the whole tree (subplans trees included)

I agree I can see usecases for both. I think the optimizer rules that handle subqueries specially currently do their own walk of the tree. For example https://github.com/apache/arrow-datafusion/blob/cc1db8a2043c73bda7adec309b42c08d88defab8/datafusion/optimizer/src/decorrelate_predicate_subquery.rs doesn't specify `apply_order` and walks down the tree




> Shall I open a PR that adds `LogicalPlan::rewriteWithSubqueries()` and `LogicalPlan::transform...WithSubqueries()`?

It seems like a good idea to me. If you you are going to do this, however, I would recommend finding some manual recursion / optimizer pass (maybe `decorrelate_predicate_subquery.rs`?) that you can use for a usecase of the new API.

Thank you again for all your help

Reply:
I've opened a PR here: https://github.com/apache/arrow-datafusion/pull/9913.
I think the new `LogicalPlan` methods like `LogicalPlan::transform_expressions()` can be useful for 2. in https://github.com/apache/arrow-datafusion/pull/9780#issuecomment-2031702263

Reply:
Update here I have created two PRs as discussed
* https://github.com/apache/arrow-datafusion/pull/9948
* https://github.com/apache/arrow-datafusion/pull/9946

Unfortunately, currently individually they either don't improve performance or actually slow things down. 

I am going to test what happens when I run them both together

Reply:
When I combined both https://github.com/apache/arrow-datafusion/pull/9948 and  https://github.com/apache/arrow-datafusion/pull/9946  tpch planning time improves by 12% (basically the same as this PR)

Thus I conclude we do not need the tree mutator API quite yet. Let's continue the discussions on those other PRs

<details><summary>Details</summary>
<p>

```
group                                         both_tests                             main                                   optimizer_tree_node2
-----                                         ----------                             ----                                   --------------------
logical_aggregate_with_join                   1.01  1203.3±30.57µs        ? ?/sec    1.00  1197.3±17.89µs        ? ?/sec    1.00  1194.4±22.14µs        ? ?/sec
logical_plan_tpcds_all                        1.00    157.6±2.32ms        ? ?/sec    1.00    157.3±3.68ms        ? ?/sec    1.00    157.6±3.67ms        ? ?/sec
logical_plan_tpch_all                         1.01     16.9±0.34ms        ? ?/sec    1.00     16.7±0.22ms        ? ?/sec    1.04     17.5±0.37ms        ? ?/sec
logical_select_all_from_1000                  1.01     19.6±0.17ms        ? ?/sec    1.00     19.4±0.17ms        ? ?/sec    1.00     19.4±0.17ms        ? ?/sec
logical_select_one_from_700                   1.01   790.8±12.36µs        ? ?/sec    1.00   784.0±12.23µs        ? ?/sec    1.00   786.1±12.87µs        ? ?/sec
logical_trivial_join_high_numbered_columns    1.00   740.8±12.72µs        ? ?/sec    1.00   737.2±28.09µs        ? ?/sec    1.00   739.5±12.25µs        ? ?/sec
logical_trivial_join_low_numbered_columns     1.00   719.9±14.19µs        ? ?/sec    1.02   731.2±11.31µs        ? ?/sec    1.00   721.4±19.28µs        ? ?/sec
physical_plan_tpcds_all                       1.00  1666.1±16.04ms        ? ?/sec    1.12  1872.6±16.17ms        ? ?/sec    1.20       2.0±0.02s        ? ?/sec
physical_plan_tpch_all                        1.00    110.2±1.86ms        ? ?/sec    1.10    121.3±1.90ms        ? ?/sec    1.21    133.3±2.21ms        ? ?/sec
physical_plan_tpch_q1                         1.00      6.2±0.11ms        ? ?/sec    1.19      7.4±0.12ms        ? ?/sec    1.23      7.6±0.07ms        ? ?/sec
physical_plan_tpch_q10                        1.00      5.1±0.07ms        ? ?/sec    1.11      5.6±0.11ms        ? ?/sec    1.20      6.1±0.10ms        ? ?/sec
physical_plan_tpch_q11                        1.00      4.5±0.09ms        ? ?/sec    1.10      5.0±0.07ms        ? ?/sec    1.20      5.4±0.10ms        ? ?/sec
physical_plan_tpch_q12                        1.00      3.6±0.06ms        ? ?/sec    1.11      4.0±0.07ms        ? ?/sec    1.21      4.4±0.11ms        ? ?/sec
physical_plan_tpch_q13                        1.00      2.4±0.04ms        ? ?/sec    1.10      2.7±0.06ms        ? ?/sec    1.18      2.9±0.08ms        ? ?/sec
physical_plan_tpch_q14                        1.00      3.1±0.05ms        ? ?/sec    1.09      3.4±0.06ms        ? ?/sec    1.19      3.7±0.09ms        ? ?/sec
physical_plan_tpch_q16                        1.00      4.5±0.06ms        ? ?/sec    1.10      4.9±0.08ms        ? ?/sec    1.24      5.6±0.10ms        ? ?/sec
physical_plan_tpch_q17                        1.00      4.3±0.07ms        ? ?/sec    1.09      4.7±0.05ms        ? ?/sec    1.21      5.2±0.10ms        ? ?/sec
physical_plan_tpch_q18                        1.00      4.6±0.09ms        ? ?/sec    1.10      5.1±0.09ms        ? ?/sec    1.22      5.6±0.10ms        ? ?/sec
physical_plan_tpch_q19                        1.00      9.0±0.11ms        ? ?/sec    1.06      9.5±0.11ms        ? ?/sec    1.23     11.1±0.14ms        ? ?/sec
physical_plan_tpch_q2                         1.00      9.6±0.12ms        ? ?/sec    1.11     10.6±0.18ms        ? ?/sec    1.23     11.7±0.14ms        ? ?/sec
physical_plan_tpch_q20                        1.00      5.5±0.08ms        ? ?/sec    1.11      6.1±0.08ms        ? ?/sec    1.24      6.9±0.14ms        ? ?/sec
physical_plan_tpch_q21                        1.00      7.6±0.11ms        ? ?/sec    1.12      8.5±0.14ms        ? ?/sec    1.22      9.3±0.13ms        ? ?/sec
physical_plan_tpch_q22                        1.00      4.1±0.10ms        ? ?/sec    1.10      4.5±0.10ms        ? ?/sec    1.23      5.0±0.12ms        ? ?/sec
physical_plan_tpch_q3                         1.00      3.6±0.07ms        ? ?/sec    1.08      3.9±0.06ms        ? ?/sec    1.19      4.3±0.11ms        ? ?/sec
physical_plan_tpch_q4                         1.00      2.7±0.06ms        ? ?/sec    1.11      3.0±0.05ms        ? ?/sec    1.21      3.2±0.11ms        ? ?/sec
physical_plan_tpch_q5                         1.00      5.3±0.08ms        ? ?/sec    1.08      5.7±0.12ms        ? ?/sec    1.17      6.2±0.08ms        ? ?/sec
physical_plan_tpch_q6                         1.00  1905.1±30.83µs        ? ?/sec    1.06      2.0±0.03ms        ? ?/sec    1.15      2.2±0.05ms        ? ?/sec
physical_plan_tpch_q7                         1.00      7.0±0.11ms        ? ?/sec    1.08      7.6±0.12ms        ? ?/sec    1.19      8.4±0.14ms        ? ?/sec
physical_plan_tpch_q8                         1.00      8.9±0.14ms        ? ?/sec    1.10      9.7±0.18ms        ? ?/sec    1.19     10.6±0.16ms        ? ?/sec
physical_plan_tpch_q9                         1.00      6.7±0.11ms        ? ?/sec    1.10      7.3±0.10ms        ? ?/sec    1.21      8.0±0.10ms        ? ?/sec
physical_select_all_from_1000                 1.00    114.1±0.77ms        ? ?/sec    1.13    129.2±1.14ms        ? ?/sec    1.12    128.3±0.79ms        ? ?/sec
physical_select_one_from_700                  1.00      3.9±0.08ms        ? ?/sec    1.03      4.1±0.05ms        ? ?/sec    1.04      4.1±0.04ms        ? ?/sec
```
</p>
</details>

Reply:
Very nice improvement! Thanks @alamb!